adapters/repos/db/vector/hnsw/hnsw_stress_test.go

```diff
@@ -0,0 +1,579 @@
+	float := math.Float32frombits(bits)
+	return float
+}
+
+func int32FromBytes(bytes []byte) int {
+	return int(binary.LittleEndian.Uint32(bytes))
+}
+
+func TestHnswStress(t *testing.T) {
+	siftFile := "datasets/siftsmall/siftsmall_base.fvecs"
+	if _, err := os.Stat(siftFile); err != nil {
+		t.Skip("Sift data needs to be present")
+	}
+	vectors := readSiftFloat(siftFile, parallelGoroutines*vectorsPerGoroutine)
+
+	t.Run("Insert and search and maybe delete", func(t *testing.T) {
+		for n := 0; n < 1; n++ { // increase if you don't want to reread SIFT for every run
+			wg := sync.WaitGroup{}
+			index := createEmptyHnswIndexForTests(t, idVector)
+			for k := 0; k < parallelGoroutines; k++ {
+				wg.Add(2)
+				goroutineIndex := k * vectorsPerGoroutine
+				go func() {
+					for i := 0; i < vectorsPerGoroutine; i++ {
+
+						err := index.Add(uint64(goroutineIndex+i), vectors[goroutineIndex+i])
+						require.Nil(t, err)
+					}
+					wg.Done()
+				}()
+
+				go func() {
+					for i := 0; i < vectorsPerGoroutine; i++ {
+						for j := 0; j < 5; j++ { // try a couple of times to delete if found
+							_, dists, err := index.SearchByVector(vectors[goroutineIndex+i], 0, nil)
+							require.Nil(t, err)
+
+							if len(dists) > 0 && dists[0] == 0 {
+								err := index.Delete(uint64(goroutineIndex + i))
+								require.Nil(t, err)
+								break
+							} else {
+								continue
+							}
+						}
+					}
+					wg.Done()
+				}()
+			}
+			wg.Wait()
+		}
+	})
+
+	t.Run("Insert and delete", func(t *testing.T) {
+		for i := 0; i < 1; i++ { // increase if you don't want to reread SIFT for every run
+			wg := sync.WaitGroup{}
+			index := createEmptyHnswIndexForTests(t, idVector)
+			for k := 0; k < parallelGoroutines; k++ {
+				wg.Add(1)
+				goroutineIndex := k * vectorsPerGoroutine
+				go func() {
+					for i := 0; i < vectorsPerGoroutine; i++ {
+
+						err := index.Add(uint64(goroutineIndex+i), vectors[goroutineIndex+i])
+						require.Nil(t, err)
+						err = index.Delete(uint64(goroutineIndex + i))
+						require.Nil(t, err)
+
+					}
+					wg.Done()
+				}()
+
+			}
+			wg.Wait()
+
+		}
+	})
+
+	t.Run("Concurrent deletes", func(t *testing.T) {
+		for i := 0; i < 10; i++ { // increase if you don't want to reread SIFT for every run
+			wg := sync.WaitGroup{}
+
+			index := createEmptyHnswIndexForTests(t, idVector)
+			deleteIds := make([]uint64, 50)
+			for j := 0; j < len(deleteIds); j++ {
+				err := index.Add(uint64(j), vectors[j])
+				require.Nil(t, err)
+				deleteIds[j] = uint64(j)
+			}
+			wg.Add(2)
+
+			go func() {
+				err := index.Delete(deleteIds[25:]...)
+				require.Nil(t, err)
+				wg.Done()
+			}()
+			go func() {
+				err := index.Delete(deleteIds[:24]...)
+				require.Nil(t, err)
+				wg.Done()
+			}()
+
+			wg.Wait()
+
+			time.Sleep(time.Microsecond * 100)
+			index.Lock()
+			require.NotNil(t, index.nodes[24])
+			index.Unlock()
+
+		}
+	})
+}
+
+func TestHnswStressRandom(t *testing.T) {
```

Comment:
I would add it as a subtest of `TestHnswStress`. Otherwise we will load the sift dataset multiple times - there you could also add the for loop around each test to be able to run it repeatedly without reloading the dataset

Reply:
It makes sense yeah